maven-extension/src/main/java/io/opentelemetry/maven/OtelTransferListener.java

```diff
@@ -0,0 +1,161 @@
+        break;
+      case GET_EXISTENCE:
+        httpRequestMethod = "HEAD";
+        break;
+      default:
+        logger.warn("OpenTelemetry: Unknown request type {}", event.getRequestType());
+        httpRequestMethod = event.getRequestType().name();
+    }
+
+    String urlTemplate =
+        event.getResource().getRepositoryUrl()
+            + "$groupId/$artifactId/$version/$artifactId-$version.$classifier";
+
+    String spanName = httpRequestMethod + " " + urlTemplate;
+
+    SpanBuilder spanBuilder =
+        this.openTelemetrySdkService
+            .getTracer()
+            .spanBuilder(spanName)
+            .setSpanKind(SpanKind.CLIENT)
+            .setAttribute(HttpAttributes.HTTP_REQUEST_METHOD, httpRequestMethod)
+            .setAttribute(
+                UrlAttributes.URL_PATH,
+                event.getResource().getRepositoryUrl() + event.getResource().getResourceName())
+            .setAttribute(UrlIncubatingAttributes.URL_TEMPLATE, urlTemplate)
+            .setAttribute(
+                MavenOtelSemanticAttributes.MAVEN_TRANSFER_TYPE, event.getRequestType().name());
+
+    repositoryUriMapping
+        .computeIfAbsent(
+            event.getResource().getRepositoryUrl(),
+            str -> {
+              try {
+                return str.isEmpty() ? Optional.empty() : Optional.of(new URI(str));
+              } catch (URISyntaxException e) {
+                return Optional.empty();
+              }
+            })
+        .ifPresent(
+            uri -> {
+              spanBuilder.setAttribute(ServerAttributes.SERVER_ADDRESS, uri.getHost());
+              if (uri.getPort() != -1) {
+                spanBuilder.setAttribute(ServerAttributes.SERVER_PORT, uri.getPort());
+              }
+              // prevent ever increasing size
+              if (repositoryUriMapping.size() > 128) {
+                repositoryUriMapping.clear();
+              }
+            });
+    spanRegistry.putSpan(spanBuilder.startSpan(), event);
+  }
+
+  @Override
+  public void transferSucceeded(TransferEvent event) {
+    logger.info("OpenTelemetry: OtelTransferListener#transferSucceeded({})", event);
+
+    Optional.ofNullable(spanRegistry.removeSpan(event))
+        .ifPresent(
+            span -> {
+              span.setStatus(StatusCode.OK);
+              finish(span, event);
+            });
+  }
+
+  @Override
+  public void transferFailed(TransferEvent event) {
+    logger.info("OpenTelemetry: OtelTransferListener#transferFailed({})", event);
+
+    Optional.ofNullable(spanRegistry.removeSpan(event)).ifPresent(span -> fail(span, event));
+  }
+
+  @Override
+  public void transferCorrupted(TransferEvent event) {
+    logger.info("OpenTelemetry: OtelTransferListener#transferCorrupted({})", event);
+
+    Optional.ofNullable(spanRegistry.removeSpan(event)).ifPresent(span -> fail(span, event));
+  }
+
+  void finish(Span span, TransferEvent event) {
+    switch (event.getRequestType()) {
```

Comment:
@smurf667 did you identify a problem in this logic? I guess that the `GET_EXISTENCE` / `HTTP HEAD` will have a transfer size of `0` but who knows.